projects/packages/backup-helper-script-manager/src/class-helper-script-manager.php

```diff
@@ -0,0 +1,584 @@
+				array( 'status' => 500 )
+			);
+		}
+		$script_body = str_replace(
+			$wp_path_marker,
+			$normalized_abspath,
+			$script_body,
+			$wp_path_marker_replacement_count
+		);
+		if ( 0 === $wp_path_marker_replacement_count ) {
+			return new \WP_Error(
+				'no_wp_path_marker',
+				"Helper script does not have the '$wp_path_marker' marker",
+				array( 'status' => 400 )
+			);
+		}
+
+		$failure_paths_and_reasons = array();
+
+		foreach ( $this->install_locations() as $directory => $url ) {
+
+			if ( is_wp_error( $url ) ) {
+				$failure_paths_and_reasons[] = "directory '$directory': " . $url->get_error_message();
+				continue;
+			}
+
+			try {
+				$installed = $this->install_to_location_or_throw( $script_body, $directory, $url );
+
+				// Always schedule a cleanup run shortly after EXPIRY_TIME.
+				\wp_schedule_single_event(
+					time() + static::EXPIRY_TIME + 60,
+					'jetpack_backup_cleanup_helper_scripts'
+				);
+
+				return array(
+					'path'    => $installed['path'],
+					'url'     => $installed['url'],
+					'abspath' => Throw_On_Errors::t_realpath( ABSPATH ),
+				);
+
+			} catch ( Exception $exception ) {
+				$failure_paths_and_reasons[] = "directory '$directory' (URL '$url'): " . $exception->getMessage();
+			}
+		}
+
+		return new \WP_Error(
+			'all_locations_failed',
+			'Unable to write the helper script to any install locations; ' .
+			'tried: ' . implode( ';', $failure_paths_and_reasons ),
+			array( 'status' => 500 )
+		);
+	}
+
+	/**
+	 * Install helper script to a directory, or throw an exception.
+	 *
+	 * @param string $script_body Helper script's body.
+	 * @param string $directory Candidate directory to create "jetpack-temp" in and write the helper script.
+	 * @param string $url Base URL that the files in a directory are expected to be available at.
+	 *
+	 * @return string[] Array with "path" (location to the installed helper script) and "url"
+	 *   (URL of the installed helper script) keys.
+	 * @throws Exception On I/O errors.
+	 */
+	protected function install_to_location_or_throw( $script_body, $directory, $url ) {
+		if ( ! Throw_On_Errors::t_is_writable( $directory ) ) {
+			throw new Exception( "Directory '$directory' is not writable" );
+		}
+
+		$temp_dir = trailingslashit( $directory ) . static::TEMP_DIRECTORY;
+
+		if ( ! Throw_On_Errors::t_is_dir( $temp_dir ) ) {
+			Throw_On_Errors::t_mkdir( $temp_dir );
+		}
+
+		$readme_path = trailingslashit( $temp_dir ) . 'README';
+		Throw_On_Errors::t_file_put_contents( $readme_path, implode( "\n\n", static::README_LINES ) );
+
+		$index_path = trailingslashit( $temp_dir ) . 'index.php';
+		Throw_On_Errors::t_file_put_contents( $index_path, static::INDEX_FILE );
+
+		$file_key  = wp_generate_password( 10, false );
+		$file_name = 'jp-helper-' . $file_key . '.php';
+		$file_path = trailingslashit( $temp_dir ) . $file_name;
+		Throw_On_Errors::t_file_put_contents( $file_path, $script_body );
```

Comment:
In the current implementation, it is validated if the helper file [already exists ](https://github.com/Automattic/jetpack/blob/739375073fc0a3976b296db736c821c6336af7bb/projects/packages/transport-helper/src/class-helper-script-manager.php#L70). Should we add that validation here too to prevent override it?

Reply:
It's very unlikely that it will already exist (1 out of 62^10 chance on case-sensitive filesystems, such as ext4 by default, or 1 out of 36^10 chance on filesystems mounted as case-insensitive), so I've omitted that check originally, but nonetheless I've readded that check in 7c9ceeebeca4ef4d1bb5cf3dd8fe77c56c3df44b.

Reply:
Excellent. Thank you!